src/solvers/cma_es.jl

```diff
@@ -0,0 +1,710 @@
+  absolute difference between function values at subsequent points
+* `tol_x`   (`1e-12`) tolerance for the `StopWhenPopulationStronglyConcentrated`, similar to
+  absolute difference between subsequent point but actually computed from distribution
+  parameters.
+
+# Output
+
+the obtained (approximate) minimizer ``p^*``.
+To obtain the whole final state of the solver, see [`get_solver_return`](@ref) for details.
+"""
+function cma_es(
+    M::AbstractManifold,
+    mco::O,
+    p_m=rand(M);
+    σ::Real=1.0,
+    λ::Int=4 + Int(floor(3 * log(manifold_dimension(M)))), # Eq. (48)
+    tol_fun::Real=1e-12,
+    tol_x::Real=1e-12,
+    stopping_criterion::StoppingCriterion=default_cma_es_stopping_criterion(
+        M, λ; tol_fun=tol_fun, tol_x=tol_x
+    ),
+    retraction_method::AbstractRetractionMethod=default_retraction_method(M, typeof(p_m)),
+    vector_transport_method::AbstractVectorTransportMethod=default_vector_transport_method(
+        M, typeof(p_m)
+    ),
+    basis::AbstractBasis=DefaultOrthonormalBasis(),
+    rng::AbstractRNG=default_rng(),
+    kwargs..., #collect rest
+) where {O<:Union{AbstractManifoldCostObjective,AbstractDecoratedManifoldObjective}}
+    dmco = decorate_objective!(M, mco; kwargs...)
+    mp = DefaultManoptProblem(M, dmco)
+    n_coords = number_of_coordinates(M, basis)
+    wp = [log((λ + 1) / 2) - log(i) for i in 1:λ] # Eq. (49)
+    μ = Int(floor(λ / 2)) # Table 1 caption
+    μ_eff = (sum(wp[1:μ])^2) / (sum(x -> x^2, wp[1:μ]))  # Table 1 caption
+    @assert μ_eff >= 1
+    @assert μ_eff <= μ
+    μ_eff⁻ = (sum(wp[(μ + 1):end])^2) / (sum(x -> x^2, wp[(μ + 1):end]))
+
+    αμ_eff⁻ = 1 + (2 * μ_eff⁻) / (μ_eff + 2) # Eq. (51)
+    α_cov = 2.0 # Eq. (57)
+    c_1 = α_cov / ((n_coords + 1.3)^2 + μ_eff) # Eq. (57)
+    c_μ = min(
+        1 - c_1,
+        α_cov * (0.25 + μ_eff + 1 / μ_eff - 2) / ((n_coords + 2)^2 + α_cov * μ_eff / 2),
+    ) # Eq. (58)
+
+    αμ⁻ = 1 + c_1 / c_μ # Eq. (50)
+    α_posdef⁻ = (1 - c_1 - c_μ) / (n_coords * c_μ) # Eq. (52)
+    w_normalization_positive = 1 / sum([wj for wj in wp if wj > 0]) # Eq. (53)
+    w_normalization_negative =
+        -min(αμ⁻, αμ_eff⁻, α_posdef⁻) / sum([wj for wj in wp if wj < 0]) # Eq. (53)
+    recombination_weights = [
+        w_i * (w_i >= 0 ? w_normalization_positive : w_normalization_negative) for w_i in wp
+    ] # Eq. (53)
+    c_m = 1.0 # Note below Eq. (9)
+
+    @assert sum(recombination_weights[1:μ]) ≈ 1
+    c_σ = (μ_eff + 2) / (n_coords + μ_eff + 5) # Eq. (55)
+    d_σ = 1 + 2 * max(0, sqrt((μ_eff - 1) / (n_coords + 1)) - 1) + c_σ # Eq. (55)
+    c_c = (4 + μ_eff / n_coords) / (n_coords + 4 + 2 * μ_eff / n_coords) # Eq. (56)
+    population = [allocate(M, p_m) for _ in 1:λ]
+    # approximation of expected value of norm of standard n_coords-variate normal distribution
+    e_mv_norm = sqrt(n_coords) * (1 - 1 / (4 * n_coords) + 1 / (21 * n_coords))
+    covm = Matrix{number_eltype(p_m)}(I, n_coords, n_coords)
+    state = CMAESState(
```

Comment:
This looks like the state could very well have a nicer constructor that itself initialises all internal fields.

Also default constructors for state take the manifold as first arg to fill in good defaults from the manifold (like the default retraction).